src/Icicle/Avalanche/Check.hs

```diff
@@ -88,7 +88,13 @@ checkStatement frag ctx stmt
                    $ checkExp frag (ctxExp ctx) x
                Let n x' <$> go stmts
 
-        ForeachInts n from to stmts
+        While t n to stmts
+         -> do to'   <- first ProgramErrorExp
+                      $ checkExp frag (ctxExp ctx) to
+
+               While t n to' <$> go stmts
```

Comment:
ah. but if the equality check is hidden in the `While` expression itself, it will not be melted. so I suppose it is better off restricting it to primitives. or maybe just restricting it to a `Bool`.